kusto/kcsb.go

```diff
@@ -11,191 +10,351 @@ import (
+		kcsb.AADUserID = value
+	}
+	if contains([]string{"password", "pwd"}, rawKey) {
+		kcsb.Password = value
+	}
+	if contains([]string{"application client id", "applicationclientid", "appclientid"}, rawKey) {
+		kcsb.ApplicationClientId = value
+	}
+	if contains([]string{"application key", "applicationkey", "appkey"}, rawKey) {
+		kcsb.ApplicationKey = value
+	}
+	if contains([]string{"application certificate", "applicationcertificate"}, rawKey) {
+		kcsb.ApplicationCertificate = value
+	}
+	if contains([]string{"application certificate thumbprint", "applicationcertificatethumbprint"}, rawKey) {
+		kcsb.ApplicationCertificateThumbprint = value
+	}
+	if contains([]string{"sendcertificatechain", "send certificate chain"}, rawKey) {
+		kcsb.SendCertificateChain = getBoolValue(value)
+	}
+	if contains([]string{"authority id", "authorityid", "authority", "tenantid", "tenant", "tid"}, rawKey) {
+		kcsb.AuthorityId = value
+	}
+	if contains([]string{"application token", "applicationtoken", "apptoken"}, rawKey) {
+		kcsb.ApplicationToken = value
+	}
+	if contains([]string{"user token", "usertoken", "usrtoken"}, rawKey) {
+		kcsb.UserToken = value
+	}
+	if contains([]string{"msi_auth"}, rawKey) {
+		kcsb.MSIAuthentication = getBoolValue(value)
+	}
+	if contains([]string{"managedserviceidentity", "managed service identity"}, rawKey) {
+		kcsb.ManagedServiceIdentity = value
+	}
+	if contains([]string{"az cli", "azcli"}, rawKey) {
+		kcsb.AZCLI = getBoolValue(value)
+	}
+	if contains([]string{"interactive login", "interactivelogin"}, rawKey) {
+		kcsb.InteractiveLogin = getBoolValue(value)
+	}
+	if contains([]string{"domain hint", "domainhint"}, rawKey) {
+		kcsb.RedirectURL = value
+	}
+}
+
+func getBoolValue(val string) bool {
+	return strings.ToLower(val) == "true"
+}
+
+/*
+   Creates new KustoConnectionStringBuilder.
+   Params takes kusto connection string connStr: string.  Kusto connection string should be of the format:
+   https://<clusterName>.kusto.windows.net;AAD User ID="user@microsoft.com";Password=P@ssWord
+   For more information please look at:
+   https://docs.microsoft.com/azure/data-explorer/kusto/api/connection-strings/kusto
+*/
+func GetConnectionStringBuilder(connStr string) connectionStringBuilder {
+	kcsb := connectionStringBuilder{}
+	if isEmpty(connStr) {
+		panic("Error : Connection string cannot be empty")
+	}
+
+	if !strings.Contains(strings.Split(connStr, ";")[0], "=") {
+		connStr = "Data Source=" + connStr
+	}
+
+	for _, kvp := range strings.Split(connStr, ";") {
+		kvparr := strings.Split(kvp, "=")
+		val := strings.Trim(kvparr[1], " ")
+		if isEmpty(val) {
+			continue
+		}
+		assignValue(&kcsb, kvparr[0], val)
+
+	}
+
+	return kcsb
+}
+
+//Creates a KustoConnection string builder that will authenticate with AAD user name and password.
+func (kcsb connectionStringBuilder) WithAadUserPassAuth(uname string, pswrd string, authorityID string) connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
```

Comment:
In hindsight , will a variadic function that takes all the parameters for the null check at once be simpler than checking each parameter ?

Reply:
Would it make sense to have *connectionStringBuilder and changing the attributes / adding attributes as opposed to taking a value and returning the value again with the mutation ?

Reply:
Because it doesn't take a pointer it doesn't mutate the provided value. It would have done it if it was a pointer.

Maybe these need to be ctors anyway, since they are mutually exclusive.